File: app/src/main/java/com/droidcode/project/WorkspaceManager.java
package com.droidcode.project;

import android.content.Context;

import com.droidcode.db.AppDatabase;
import com.droidcode.db.WorkspaceDao;
import com.droidcode.db.WorkspaceEntity;
import com.droidcode.filesystem.FileNode;
import com.droidcode.filesystem.LocalFileSystem;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class WorkspaceManager {

    private static volatile WorkspaceManager INSTANCE;

    private Project currentProject;
    private final LocalFileSystem fileSystem;

    private WorkspaceManager() {
        this.fileSystem = LocalFileSystem.getInstance();
    }

    public static WorkspaceManager getInstance() {
        if (INSTANCE == null) {
            synchronized (WorkspaceManager.class) {
                if (INSTANCE == null) {
                    INSTANCE = new WorkspaceManager();
                }
            }
        }
        return INSTANCE;
    }

    public Project getCurrentProject() {
        return currentProject;
    }

    public boolean hasOpenWorkspace() {
        return currentProject != null && new File(currentProject.getPath()).exists();
    }

    public void openWorkspace(Context context, File directory, String projectType) {
        if (directory == null || !directory.exists() || !directory.isDirectory()) {
            throw new IllegalArgumentException("Target directory does not exist or is invalid");
        }

        Project project = Project.fromDirectory(directory, projectType);
        if (project != null) {
            this.currentProject = project;

            // Save or update in database
            AppDatabase db = AppDatabase.getInstance(context);
            WorkspaceDao dao = db.workspaceDao();
            dao.insertWorkspace(new WorkspaceEntity(
                    project.getPath(),
                    project.getName(),
                    project.getType(),
                    project.getLastOpenedAt(),
                    project.isGitRepo()
            ));
        }
    }

    public void closeWorkspace() {
        this.currentProject = null;
    }

    public List<FileNode> getWorkspaceFileTree() {
        if (!hasOpenWorkspace()) {
            return new ArrayList<>();
        }
        return fileSystem.listDirectoryRecursive(currentProject.getDirectory());
    }

    public List<WorkspaceEntity> getRecentWorkspaces(Context context) {
        AppDatabase db = AppDatabase.getInstance(context);
        List<WorkspaceEntity> list = db.workspaceDao().getAllWorkspacesSync();
        return list != null ? list : new ArrayList<>();
    }

    public void removeWorkspace(Context context, String path) {
        if (path != null && context != null) {
            AppDatabase db = AppDatabase.getInstance(context);
            db.workspaceDao().deleteWorkspaceByPath(path);
        }
    }
}

File: app/src/main/java/com/droidcode/project/WorkspaceManager.kt
package com.droidcode.project

import android.content.Context
import com.droidcode.db.AppDatabase
import com.droidcode.db.WorkspaceEntity
import com.droidcode.filesystem.FileNode
import com.droidcode.filesystem.LocalFileSystem
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.withContext
import java.io.File
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class WorkspaceManager @Inject constructor() {

    @Volatile
    private var _currentProject: Project? = null

    val currentProject: Project?
        get() = _currentProject

    private val _currentProjectFlow = MutableStateFlow<Project?>(null)
    val currentProjectFlow: StateFlow<Project?> = _currentProjectFlow.asStateFlow()

    private val _treeVersion = MutableStateFlow(0)
    val treeVersion: StateFlow<Int> = _treeVersion.asStateFlow()

    fun notifyTreeChanged() {
        _treeVersion.value += 1
    }

    private val fileSystem: LocalFileSystem = LocalFileSystem.getInstance()

    val hasOpenWorkspace: Boolean
        get() {
            val proj = _currentProject ?: return false
            return File(proj.path).exists()
        }

    fun hasOpenWorkspace(): Boolean = hasOpenWorkspace

    suspend fun openWorkspace(context: Context, directory: File?, projectType: String?) = withContext(Dispatchers.IO) {
        if (directory == null || !directory.exists() || !directory.isDirectory) {
            throw IllegalArgumentException("Target directory does not exist or is invalid")
        }

        val project = Project.fromDirectory(directory, projectType)
        if (project != null) {
            _currentProject = project
            _currentProjectFlow.value = project
            _treeVersion.value += 1

            val db = AppDatabase.getInstance(context)
            db.workspaceDao().insertWorkspace(
                WorkspaceEntity(
                    project.path,
                    project.name,
                    project.type,
                    project.lastOpenedAt,
                    project.isGitRepo
                )
            )
        }
    }

    fun closeWorkspace() {
        _currentProject = null
        _currentProjectFlow.value = null
        _treeVersion.value += 1
    }

    val workspaceFileTree: List<FileNode>
        get() {
            val proj = _currentProject ?: return emptyList()
            if (!File(proj.path).exists()) return emptyList()
            return fileSystem.listDirectoryRecursive(proj.directory)
        }

    fun getRecentWorkspacesFlow(context: Context): Flow<List<WorkspaceEntity>> {
        val db = AppDatabase.getInstance(context)
        return db.workspaceDao().getAllWorkspacesFlow()
    }

    suspend fun getRecentWorkspaces(context: Context): List<WorkspaceEntity> = withContext(Dispatchers.IO) {
        val db = AppDatabase.getInstance(context)
        db.workspaceDao().getAllWorkspaces()
    }

    suspend fun removeWorkspace(context: Context, path: String?) = withContext(Dispatchers.IO) {
        if (path != null) {
            val db = AppDatabase.getInstance(context)
            db.workspaceDao().deleteWorkspaceByPath(path)
        }
    }

    companion object {
        @Volatile
        private var INSTANCE: WorkspaceManager? = null

        @JvmStatic
        fun getInstance(): WorkspaceManager {
            return INSTANCE ?: synchronized(WorkspaceManager::class.java) {
                INSTANCE ?: WorkspaceManager().also { INSTANCE = it }
            }
        }
    }
}
